### PR TITLE
upgrade: introduce "permanent" upgrades

### LINKED ISSUE #73813: migration/startupmigrations: remove all startupmigrations
**Is your feature request related to a problem? Please describe.**

Before long-running migrations, there were migrations which ran at startup. We haven't really leveraged these in a while (none since 20.2). There are, however, "permanent" migrations which set up some initial cluster state. We should get rid of that concept and all the complexity that comes with it.

**Describe the solution you'd like**
Move any permanent migrations into bootstrap. Some of the remaining seem to not really be permanent. 

#### Remaining permanent migrations:

* [`optIntoDiagnosticsReporting`](https://github.com/cockroachdb/cockroach/blob/42055b0c16528d0e7e2e78e255430ffec158d205/pkg/startupmigrations/migrations.go#L95-L99)
* [`populateVersionSetting`](https://github.com/cockroachdb/cockroach/blob/42055b0c16528d0e7e2e78e255430ffec158d205/pkg/startupmigrations/migrations.go#L108-L113)
* [`addRootUser`](https://github.com/cockroachdb/cockroach/blob/42055b0c16528d0e7e2e78e255430ffec158d205/pkg/startupmigrations/migrations.go#L119-L123)
* [`addAdminRole`](https://github.com/cockroachdb/cockroach/blob/42055b0c16528d0e7e2e78e255430ffec158d205/pkg/startupmigrations/migrations.go#L138-L142)
* [`addRootToAdminRole`](https://github.com/cockroachdb/cockroach/blob/42055b0c16528d0e7e2e78e255430ffec158d205/pkg/startupmigrations/migrations.go#L147-L151)
* [`initializeClusterSecret`](https://github.com/cockroachdb/cockroach/blob/42055b0c16528d0e7e2e78e255430ffec158d205/pkg/startupmigrations/migrations.go#L168-L172)
* [`disallowPublicUserOrRole`](https://github.com/cockroachdb/cockroach/blob/42055b0c16528d0e7e2e78e255430ffec158d205/pkg/startupmigrations/migrations.go#L182-L187)
* [`createDefaultDatabases`](https://github.com/cockroachdb/cockroach/blob/42055b0c16528d0e7e2e78e255430ffec158d205/pkg/startupmigrations/migrations.go#L188-L194)
* [`retireOldTsPurgeIntervalSettings`](https://github.com/cockroachdb/cockroach/blob/42055b0c16528d0e7e2e78e255430ffec158d205/pkg/startupmigrations/migrations.go#L219-L225)
* [`updateSystemLocationData`](https://github.com/cockroachdb/cockroach/blob/42055b0c16528d0e7e2e78e255430ffec158d205/pkg/startupmigrations/migrations.go#L226-L230)
* [`extendCreateRoleWithCreateLogin`](https://github.com/cockroachdb/cockroach/blob/42055b0c16528d0e7e2e78e255430ffec158d205/pkg/startupmigrations/migrations.go#L294-L298)

**Describe alternatives you've considered**
We could not do this. An advantage to startup migrations is that they get to leverage the running system and, in doing so, may gain benefit from more code reuse. 

**Additional context**

The pressing migration which motivated this was `createDefaultDatabases`. 

Jira issue: CRDB-11759

Epic CRDB-18596

### PR DESCRIPTION
**Pull request by andreimatei**
This patch introduces "permanent" upgrades - a type of upgrade that is
tied to a particular cluster version (just like the existing upgrades)
but that runs regardless of the version at which the cluster was
bootstrapped (in contrast with the existing upgrades that are not run
when they're associated with a cluster version <= the bootstrap
version). These upgrades are called "permanent" because they cannot be
deleted from the codebase at a later point, in contrast with the others
that are deleted once the version they're tied drops below
BinaryMinSupportedVersion).

Existing upgrades are explicitly or implicitly baked into the bootstrap
image of the binary that introduced them. For example, an upgrade that
creates a system table is only run when upgrading an existing,
older-version, cluster to the new version; it does not run for a cluster
bootstrapped by the binary that introduced the upgrade because the
respective system tables are also included in the bootstrap metadata.
For some upcoming upgrades, though, including them in the bootstrap
image is difficult. For example, creating a job record at bootstrap
time is proving to be difficult (the system.jobs table has indexes, so
you want to insert into it through SQL because figuring out the kv's for
a row is tedious, etc). This is where these new permanent upgrades come
in.

These permanent upgrades replace the `startupmigrations` that don't have
the `includedInBootstrap` field set. All such startupmigrations have
been copied over as upgrades. None of the current `startupmigrations`
have `includedInBootstrap` set (except one but that's dummy one since
the actual migration code has been deleted), so the startupmigrations
package is now deleted. That's a good thing - we had one too many
migrations frameworks.

These permanent upgrades, though, do not have exactly the same semantics
as the startupmigrations they replace. To the extent that there is a
difference, the new semantics are considered more desirable:
- startupmigrations run when a node that has the code for a particular
  migration startups up for the first time. In other words, the
  startupmigrations were not associated with a cluster version; they were
  associated with a binary version. Migrations can run while old-version
  nodes are still around.  This means that one cannot add a
  migration that is a problem for old nodes - e.g. a migration creating a
  job of a type that the old version wouldn't recognize.
- upgrades are tied to a cluster version - they only run when the
  cluster's active version moves past the upgrade's version. This stays
  the case for the new permanent migrations too, so a v2 node will not
  immediately run the permant migrations introduced since v1 when it joins
  a v1 cluster. Instead, the migrations will run when the cluster version
  is bumped. As such, the migrations can be backwards incompatible.

startupmigrations do arguably have a property that can be desirable:
when there are no backwards compatibility issues, the v2 node can rely
on the effects of the startupmigrations it knows about regardless of the
cluster version. In contrast, with upgrades, not only is a node unable
to simply assume that a particular upgrade has run during startup, but,
more than that, a node is not even able to look at a version gate during
the startup sequence in order to determine whether a particular upgrade
has run or not (because, in clusters that are bootstrapped at v2, the
active cluster version starts as v2 even before the upgrades run). This
is a fact of life for existing upgrades, and now becomes a fact of life
for permanent upgrades too. However, by the time user SQL traffic is
admitted on a node, the node can rely on version gates to correspond to
migrations that have run.

After thinking about it, this possible advantage of startupmigrations
doesn't seem too useful and so it's not reason enough to keep the
startupmigrations machinery around.

Since the relevant startupmigrations have been moved over to upgrades,
and the two libraries use different methods for not running the same
migration twice, a 23.1 node that comes up in a 22.2 cluster will re-run
the several permanent upgrades in question, even though they had already
run as startupmigrations. This is OK since both startupmigrations and
upgrades are idempotent. None of the current permanent upgrades are too
expensive.

Closes https://github.com/cockroachdb/cockroach/issues/73813

Release note: None
Epic: None

### CODE REVIEW

**Comment by andreimatei** [comment]
cc @zachlite

**Comment by miretskiy** [comment]
cc @jayshrivastava

**Review by ajwerner — COMMENTED** [review]
Want to peel the first commit off into a separate PR? This post is just about the first one.

> Reviewed 9 of 19 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-:-NGUPJeB2PKS_LxERU54:bum2hag)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @miretskiy)

___
*[`pkg/roachpb/version.go` line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NGUOppE79aldnrNq1Rn:-NGUOppF1Q20u6H9jxm9:b5zwp37) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d27c8e29798717fe667cf3d7b4a1e9ad3c6216/pkg/roachpb/version.go#L73)):*
> ```Go
> 		return v.String()
> 	}
> 	return redact.Sprintf("%s%s", v.String(), "(fence)").StripMarkers()
> ```

why is this not `fmt.Sprintf("%v (fence)", v)`

**Review by postamar — COMMENTED** [review]
Thanks for doing this. This feels like a good change, I mostly have quibbles about naming things. I would prefer if this new functionality was exercised somehow to validate that it works as intended before approving it, though. 

For this purpose would it be possible to port one or several existing system table data insertions in `bootstrap` into one of these new migrations? Assigning them the minimum cluster version value should ensure that they always run. For instance I have in mind `addSystemTenantEntry` which populates `system.tenants`, which should result in noisy failures if there's bugs.

**Comment by andreimatei** [comment]
Not quite ready for another look, but soon!

**Comment by dt** [comment]
First off, my apologies for rolling in late here after there has obviously already been a lot of time and effort invested.

I'm a little concerned about how much this approach complicates the story of a simple linear line of versions and upgrade migrations being how we move along that line.

First, some background, separate from this work, it has been becoming clearer lately, particularly in the context of multi-tenant clusters, that we probably need to change how we bootstrap. Specifically, we probably need to bootstrap a created cluster - in particular a created _secondary tenant cluster_ to the minimumBinaryVersion cluster version and associated system schema/state, and then run upgrade migrations to get from there to our desired state/version. 

A big reason this change in approach, versus bootstrapping directly to the final version, seems likely to be forced on us, is tenant creation in a mixed version host cluster: a) we have a hard rule that the tenant cluster version must not exceed the host cluster cluster version b) the initial boostrap state and cluster version is determined by the binary that run it. Today, that binary bootstraps straight to its final version, regardless of the cluster version, but that means if you add one new binary to a host cluster -- but do not yet advance the host cluster version -- if that binary happens to handle a CREATE TENANT call, it'd bootstrap that tenant to a version above the host cluster version, violating our hard rule.

So it seems likely that we need to change how we bootstrap, to instead only "flatten" system state changes into to the bootstrap schema when its associated upgrade migration drops below minBinary and is deleted, and run the upgrade migrations after bootstrapping a new cluster.

Now assuming we make that change, what does that mean for this? Well, I think it could mean we have a much simpler option for "I want to make sure this gets run in every cluster" than is sketched here: we can just use a standard, run of the mill upgrade migration which we "rebase" any time it falls below minBinary, meaning it is always run, and then any piece of code that wants to e.g. create a job or a schedule or a default user or whatever can simply register itself in this migration.

WDYT?

**Comment by andreimatei** [comment]
> Now with that as background, assuming we make those changes, what does that mean for this work? Well, I think it could mean we have a much simpler option for "I want to make sure this gets run in every cluster" than is sketched here: we can just use a standard, run of the mill upgrade migration which we "rebase" any time it falls below minBinary, meaning it is always run, and then any piece of code that wants to e.g. create a job or a schedule or a default user or whatever can simply register itself in this migration.

How do you envision this "rebasing" happening? Or, rather, doesn't this patch do what you want? This patch makes it possible to have initialization that's not baked into the bootstrap image (i.e. `MakeMetadataSchema()`), which I think is what you're talking about. What did you mean by "I'm a little concerned about how much this approach complicates the story of a simple linear line of versions and upgrade migrations being how we move along that line." more concretely?

It seems to me that, whether or not we bootstrap at the `minBinary` version or at the `binaryVersion`, this patch remains good - the "permanent upgrades" are the ones that have not been folded into any image, so they always get run. As opposed to the non-permanent upgrades, which only get run for versions above the bootstrap version.

**Comment by dt** [comment]
> How do you envision this "rebasing" happening? 
When minBinary would advance past a never-baked-in migration, we'd simply renumber it to remain above minBinary. But now that I think about it, I'm not even sure we need to do that, if we just bootstrapped to _below_ minBinary, specifically to the earliest version in our list, and then migrated our way up. Anything not permeant below minBinary (which is now a point somewhere in the middle of the list) will have been baked in and removed, so the prefix of the line before minBinary is just the "permeant" ones at that point.

> Or, rather, doesn't this patch do what you want?

I guess I'm not sure we need to encode which are permeant or not, or have the distinct "run the permeant ones" phase? If we just had them all as regular ol upgrade migrations, but ensured new clusters initialized to a version below them, then they run just like any upgrade migrations, without needing to make them special at runtime. The only thing special about them is that we don't delete them.

**Comment by andreimatei** [comment]
Ah, so I think you're saying is that, if we bootstrapped at `minBinary` instead of bootstrapping at `binaryVersion`, then we could delete all migrations that we no longer need. Whereas today the situation is muddier - we need to keep migrations in the code base that run some times but don't run other times. 
Today, let's say we have a binary with `minBinaryVersion = 5, binaryVersion = 10`. This binary needs to contain all the permanent upgrades for versions `1,2,3,4,5`, and also permanent and non-permanent upgrades for versions `6,7,8,9,10`. If a node running this binary is performing the bootstrap, then it needs to run all the permanent upgrades `<= 10` (so, it must exclude the non-permanent ones in `[6,10]`). That's why we need to distinguish the permanent from non-permanent ones.

In contrast, tomorrow, if we bootstrap at `minBinaryVersion`, then the bootstrapping node needs to run the permanent upgrades in `[`1,5]` (and nothing in `[6-10]`). Later, when `set cluster version` is set, it will need to run permanent and non-permanent upgrades in `[6,10]`. Since the binary does not need to contain non-permanent upgrades <= 5, we don't need to discriminate between permanent and non-permanent upgrades - which is your point.

Did I get it right?

**Review by andreimatei — COMMENTED** [review]
First two commits are #92157.

@dt , assuming that I got the above right, I agree with you that, if we were to bootstrap at `BinaryMinSupportedVersion`, then it seems we wouldn't need the permanent/non-permanent demarcation between upgrades. In my opinion, that alone is not reason enough to change the way we bootstrap - but you say there's multi-tenant reasons to do so.
Still, what do you think of me merging this anyway? If we change the bootstrap, then we can remove the `permanent` bit. I'd rather not look into changing the bootstrap myself.
This patch does reasonable things, I think, besides that `permanent` bit:
- it deletes the `startupmigrations` pkg, transferring all those migrations to the newer and hipper `upgrades` pkg, so that we don't live with two overlapping libraries. Things were quite confusing with both of them.
- it makes the `upgrade` code able to use a bounded-staleness query to figure out if upgrades need to run. This is needed anyway if the bootstrap changes, and upgrades need to be checked on more regular code paths (i.e. on general node startup). (Doing this what quite the rabbit hole, btw, culminating in #92157).
- it does various things to the `upgrade` code so that it plays better with tests who muck with the list of upgrades to run at startup, or with the jobs infrastructure. Among other things, it teaches the `upgrade` code to run optionally without using jobs, as these jobs get in the way of sufficiently many tests. This stuff I think is again useful with the bootstrap change.

The reason this patch came to be, btw, is because several people need to create "static" jobs or schedules to do background cluster work. And different people were doing this in different ways, some of which I though were chaotic. So I want us to have a clear framework for creating these jobs, but neither `upgrade` nor `startupmigrations` was good enough to handle it before.

WDYT?

If you're generally cool with the patch, you might want to bikeshed my introduction of `VPrimordial` versions; I'm happy to do that. Then again, if this is subject to change with the bootstrap, maybe you don't care to.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-:-NHQAFea3A7l99hE1PcU:bqq7695)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, @msbutler, @postamar, @renatolabs, and @smg260)

___
*[`pkg/server/server.go` line 1719 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NGT5tHqo_aMUlwphIBq-r2-1719:-NGXL5Sn4n6v5Wv_OoPw:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e863abf03188a7b2c52808d1b275215b1bfd7ab1/pkg/server/server.go#L1719)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Did you mean `startupmigrations and permanent upgrades`?
</blockquote></details>

done
___
*[`pkg/server/server_sql.go` line 201 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NGT5tHqo_aMUlwphIBr-r2-201:-NGXLOSl-fIKOoxOg7Gw:b-alyyji) ([raw file](https://github.com/cockroachdb/cockroach/blob/e863abf03188a7b2c52808d1b275215b1bfd7ab1/pkg/server/server_sql.go#L201)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

uber nit: naming consistency, consider renaming the package to upgradesmanager
</blockquote></details>

renamed the field to `upgradeManager` instead
___
*[`pkg/server/server_sql.go` line 1464 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NGT5tHqo_aMUlwphIBr-r2-1464:-NGXLYnh6t6l5vKCABis:b-u4q2t2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e863abf03188a7b2c52808d1b275215b1bfd7ab1/pkg/server/server_sql.go#L1464)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Perhaps I'm reading too much into this: the use of quotes around "permanent" suggests some discomfort with the choice of word to qualify this new type of upgrade. Indeed, all upgrades are permanent in their effects, it's not like you can roll any of them back.

If you're open to naming suggestions, how about describing them as "unconditional upgrades"? Still, that's not quite accurate either, they are also conditional:
- an upgrade of this new kind might run only if its version <= current active cluster version;
- an upgrade of the already-existing kind might run only if its version > current active cluster version.

Perhaps consider naming each kind along the lines of "pre-upgrades" and "post-upgrades", respectively? That might be too contrived and unclear, I don't know. Food for thought.
</blockquote></details>

"permanent" was a pre-existing name for this sort of thing, as @ajwerner showed me.  It's used in comments [like this](https://github.com/cockroachdb/cockroach/blob/e863abf03188a7b2c52808d1b275215b1bfd7ab1/pkg/startupmigrations/migrations.go#L92).
I don't like the name very much, but I don't like your suggestions better...
___
*[`pkg/roachpb/version.go` line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NGUOppE79aldnrNq1Rn:-NGXKybl51spLrmS6DvE:brhj2zv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3d27c8e29798717fe667cf3d7b4a1e9ad3c6216/pkg/roachpb/version.go#L73)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why is this not `fmt.Sprintf("%v (fence)", v)`
</blockquote></details>

resolved elsewhere

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91627)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, @msbutler, @postamar, @renatolabs, and @smg260)

___
*[`-- commits` line 122 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NHQJ2e-6jtcq3MDyvIu):*
@postamar , moving this thread here

> Thanks for doing this. This feels like a good change, I mostly have quibbles about naming things. I would prefer if this new functionality was exercised somehow to validate that it works as intended before approving it, though.

> For this purpose would it be possible to port one or several existing system table data insertions in bootstrap into one of these new migrations? Assigning them the minimum cluster version value should ensure that they always run. For instance I have in mind `addSystemTenantEntry` which populates system.tenants, which should result in noisy failures if there's bugs.


I think taking moving things out of the actual cluster bootstrap (i.e. `MakeMetadataSchema`) and into an upgrade would be a mistake; I think that bootstrap is preferable whenever possible ("static" code versus jobs, etc). So I wouldn't move `addSystemTenantEntry`.
However, I've deleted `startupmigrations` and moved all of them over to `upgrade`, as permanent migrations. So the code is exercised now.

File: pkg/upgrade/upgrade.go [hunk 1]
```diff
@@ -25,72 +25,44 @@ import (
 	"fmt"
 
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/upgrade/upgradebase"
 )
 
-// Upgrade defines a program to be executed once every node in the cluster is
```
Comment (ajwerner) [inline pkg/upgrade/upgrade.go:30]: This seems backwards. Let's leave this in `upgrade` and move the dependencies and structs to a subpackage. It's not used externally.

File: pkg/upgrade/upgradebase/migrations_table.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+)
+
+// MarkMigrationCompleted inserts a row in system.migrations.
+func MarkMigrationCompleted(
```
Comment (ajwerner) [inline pkg/upgrade/upgradebase/migrations_table.go:22]: nit: stick this in its own package

File: pkg/sql/set_cluster_setting.go [hunk 3]
```diff
@@ -618,7 +618,7 @@ func runMigrationsAndUpgradeVersion(
 	prev tree.Datum,
 	value tree.Datum,
 	updateVersionSystemSetting UpdateVersionSystemSettingHook,
-) error {
+) error { // _err
```
Comment (ajwerner) [inline pkg/sql/set_cluster_setting.go:621]: ?

File: pkg/upgrade/upgrades/upgrades.go [hunk 4]
```diff
@@ -185,3 +249,254 @@ func init() {
 func toCV(key clusterversion.Key) roachpb.Version {
 	return clusterversion.ByKey(key)
 }
+
+func initializeClusterSecret(
```
Comment (ajwerner) [inline pkg/upgrade/upgrades/upgrades.go:253]: nit: put these in a file

File: pkg/upgrade/upgrademanager/manager.go [hunk 5]
```diff
@@ -148,8 +154,108 @@ func safeToUpgradeTenant(
...
+		// did not actually mean to inhibit running these. So we'll run them anyway,
+		// without using jobs, so that the side effect of the migrations are
+		// minimized and tests continue to be happy as they were before the
+		// permanent migrations were introduced.
+		err := m.runPermanentMigrationsWithoutJobsForTests(ctx, user)
+		if err != nil {
+			return err
+		}
+
+		return nil
```
Comment (ajwerner) [inline pkg/upgrade/upgrademanager/manager.go:199]: nit: `return m.runPermanentMigrationsWithoutJobsForTests(ctx, user)`

**Review by ajwerner — COMMENTED** [review]
This works for me. Thanks for doing it.

**Review by andreimatei — COMMENTED** [review]
> This seems backwards. Let's leave this in upgrade and move the dependencies and structs to a subpackage. It's not used externally.

Discussed offline and left it alone.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-:-NHWdJRQDVH3NRiN9NIa:b70kgj3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, @msbutler, @postamar, @renatolabs, and @smg260)

___
*[`pkg/upgrade/upgrademanager/manager.go` line 199 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NGT5tHp9Q2_YzZDhBvD-r9-199:-NHW4ncq6TVIOmnMhMkR:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8157e597f90999f46961c466778b595ce23b0e90/pkg/upgrade/upgrademanager/manager.go#L199)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: `return m.runPermanentMigrationsWithoutJobsForTests(ctx, user)`
</blockquote></details>

done
___
*[`pkg/upgrade/upgrades/upgrades.go` line 253 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NGT5tHp9Q2_YzZDhBvH-r9-253:-NHWd9al5oIR4bDJexQ6:bocs3o7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8157e597f90999f46961c466778b595ce23b0e90/pkg/upgrade/upgrades/upgrades.go#L253)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: put these in a file
</blockquote></details>

move to `permanent_upgrades.go`
___
*[`pkg/sql/set_cluster_setting.go` line 621 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NHRt_DHli72e76P2Dyj-r9-621:-NHW3Pu1Bd_aHDQ1doyL:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8157e597f90999f46961c466778b595ce23b0e90/pkg/sql/set_cluster_setting.go#L621)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

?
</blockquote></details>

done
___
*[`pkg/upgrade/upgradebase/migrations_table.go` line 22 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NGxou8dR_AVrQj6ijlE-r9-22:-NHW4cEt3A8hRRhalg2S:b-w5jlpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8157e597f90999f46961c466778b595ce23b0e90/pkg/upgrade/upgradebase/migrations_table.go#L22)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: stick this in its own package 
</blockquote></details>

done; also moved `CheckIfMigrationCompleted`.

**Review by ajwerner — APPROVED** [review]
:lgtm:

Needs a `./dev gen bazel --short`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-:-NHWfGQm6mMW8ypCbmn5:b4kx7sb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @miretskiy, @msbutler, @postamar, @renatolabs, and @smg260)

___
*[`pkg/upgrade/upgrades/permanent_upgrades.go` line 30 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NHWfT2J22bvDdk7wYu-:-NHWfT2J22bvDdk7wYu0:b1nnhdv) ([raw file](https://github.com/cockroachdb/cockroach/blob/921382d3229b3a64f3d9f3ca55ecddf2a8b30ae6/pkg/upgrade/upgrades/permanent_upgrades.go#L30)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
> 	"github.com/cockroachdb/cockroach/pkg/util/retry"
> 	"github.com/pkg/errors"
> ```

wrong errors

**Comment by dt** [comment]
I don't think I or anyone else is going to do the bootstrap to min then run (at least some) migrations version of init right away, so if you're actively blocked, I guess what you have here is probably as good as we're going to get in the short term, so I withdraw my suggestion to take an alternative approach. I'll bow out here and leave the rest to your existing reviewers

**Review by andreimatei — COMMENTED** [review]
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-:-NH_YjDQByFAwf1A4Frc:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @miretskiy, @msbutler, @postamar, @renatolabs, and @smg260)

___
*[`pkg/upgrade/upgrades/permanent_upgrades.go` line 30 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/91627#-NHWfT2J22bvDdk7wYu-:-NH_OR1P-kWr7E4FRhJD:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/921382d3229b3a64f3d9f3ca55ecddf2a8b30ae6/pkg/upgrade/upgrades/permanent_upgrades.go#L30)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

wrong errors
</blockquote></details>

done

**Pull request merged**